pkg/fqdn/dnsproxy/udp.go

```diff
@@ -55,37 +60,41 @@ type sessionUDP struct {
 	oob   []byte
 }
 
-var ciliumSessionUDPFactory = &sessionUDPFactory{}
-
 var rawconn4 *net.IPConn // raw socket for sending IPv4
 var rawconn6 *net.IPConn // raw socket for sending IPv6
 
 // Set the socket options needed for tranparent proxying for the listening socket
 // IP(V6)_TRANSPARENT allows socket to receive packets with any destination address/port
 // IP(V6)_RECVORIGDSTADDR tells the kernel to pass the original destination address/port on recvmsg
 // The socket may be receiving both IPv4 and IPv6 data, so set both options, if possible.
-func transparentSetsockopt(fd int) error {
-	err6 := unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_TRANSPARENT, 1)
-	if err6 == nil {
-		err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_RECVORIGDSTADDR, 1)
-	}
-	err4 := unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_TRANSPARENT, 1)
-	if err4 == nil {
-		err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_RECVORIGDSTADDR, 1)
+func transparentSetsockopt(fd int, ipv4, ipv6 bool) error {
+	var err4, err6 error
+	if ipv6 {
+		err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_TRANSPARENT, 1)
+		if err6 == nil {
+			err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_RECVORIGDSTADDR, 1)
+		}
 	}
-	// Only error out if both IPv4 and IPv6 fail
-	if err4 != nil && err6 != nil {
-		return err4
+	if ipv4 {
+		err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_TRANSPARENT, 1)
+		if err4 == nil {
+			err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_RECVORIGDSTADDR, 1)
+		}
+		// Only error out if both IPv4 and IPv6 fail, or if either fails and the
+		// other is disabled (the !ip* bits).
+		if (!ipv4 || err4 != nil) && (!ipv6 || err6 != nil) {
```

Comment:
I think this logic attempted to account for the case where ipv6 wasn't enabled, by only returning err4. Now that either type can be enabled or disabled, does this actually need to unconditionally return err6 above and err4 here?

Reply:
This would return a nil error in the case of !ipv4 but an error from ipv6, though.

Reply:
How so? I see it breaking down as below, but I think I will just make this error out if either has an error and is enabled. There is too much subtlety here.
```
(!ipv4 || err4 != nil) && (!ipv6 || err6 != nil)
(true || err4 != nil) && (!ipv6 || err6 != nil) <-- ipv4 is disabled, !ipv4 is true
(true) && (!ipv6 || err6 != nil) <-- (!ipv4 || err4 != nil) is also true because of the ||
(true) && (false || err6 != nil) <-- ipv6 is enabled, !ipv6 is false
(true) && (false || true) <-- let's say err6 != nil, hence true
(true) && (true) <-- (!ipv6 || err6 != nil) is now true
so true
```

Reply:
Oops, you are right 🤦‍♂